README.md
# Transparent Image Button  

[![pub.dev](https://img.shields.io/pub/v/transparent_image_button.svg)](https://pub.dev/packages/transparent_image_button)  [![Build Status](https://travis-ci.com/ShadyBoshra2012/flutter_transparent_image_button.svg?branch=master)](https://travis-ci.com/ShadyBoshra2012/flutter_transparent_image_button) [![Codemagic build status](https://api.codemagic.io/apps/5d4afeeb36524a00189ef782/5d4afeeb36524a00189ef781/status_badge.svg)](https://codemagic.io/apps/5d4afeeb36524a00189ef782/5d4afeeb36524a00189ef781/latest_build) [![Donate Paypal](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://paypal.me/ShadyBoshra2012) [![GitHub Follow](https://img.shields.io/github/followers/ShadyBoshra2012.svg?style=social&label=Follow)](https://github.com/ShadyBoshra2012)

## Explain  
  
This package is used to click on **Transparent Image** as a Gesture, but it's `OnTap` function works only when it was clicked on the image or photo itself not on the transparent parts.  
  
## How to use ?  
  
```
TransparentImageButton.assets(  
	 "assets/images/egg.png", 
	 width: 200, 
	 opacityThreshold: 0.8,
         onTapInside: () => print("You tapped the image."),
         onTapOutside: () => print("You tapped outside the image."),
         onHoverInside: () => print("You're hovering inside the image."),
         onHoverOutside: () => print("You're hovering outside the image."),
),
```

## Things ToDO

- [x]  Image from Assets.
- [ ]  Image from Network.

## The idea  
  The idea came to me from this [question](https://stackoverflow.com/questions/57374066/button-with-image-with-transparent-background) on Stackoverflow.  
  
## Sources that helped me  
  - [How to Calculate the Coordinates of touch in Flutter?](https://stackoverflow.com/questions/46560982/how-to-calculate-the-coordinates-of-touch-in-flutter)  
- [Get the pixel on which I clicked](https://stackoverflow.com/questions/56478321/get-the-pixel-on-which-i-clicked/56488053#56488053)  
- [roipeker](https://gist.github.com/roipeker) [gist file](https://gist.github.com/roipeker/9315aa25301f5c0362caaebd15876c2f)


## Contributors (Thanks for your GREAT work ❤️)

[Jordan](https://github.com/APartOfMe1): Added QoL improvements for Flutter web.

## Issues or Contributions

I am very welcoming any issues or contribution you can help me with.

## License

MIT: [https://mit-license.org](https://mit-license.org). 

Copyright (c) 2019 Shady Boshra. All rights reserved.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

## Find me on Stackoverflow

<a href="https://stackoverflow.com/users/2076880/shady-boshra"><img src="https://stackoverflow.com/users/flair/2076880.png" width="208" height="58" alt="profile for Shady Boshra at Stack Overflow, Q&amp;A for professional and enthusiast programmers" title="profile for Shady Boshra at Stack Overflow, Q&amp;A for professional and enthusiast programmers"></a>


CHANGELOG.md
## [1.0.0]

* Add QoL improvements for Flutter web.

## [0.2.0]

* Migrate Null-Safety.

## [0.1.1+1].- Update README file

* Update README file.

## [0.1.1].- Update README file

* Update README file.

## [0.1.0] - First pre-release version of Package.

* This First pre-release version.

## [0.0.1] - First beta version of Package.

* This beta version of package.


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter_test/flutter_test.dart';

import '../lib/main.dart';


void main() {
  testWidgets('test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());
  });
}


lib/transparent_image_button.dart
library transparent_image_button;

import 'dart:async';
import 'dart:typed_data';

import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'package:flutter/services.dart';
import 'package:image/image.dart' as img;

class TransparentImageButton extends StatefulWidget {
  final String imagePath;
  final Function? onTapInside;
  final Function? onTapOutside;
  final Function? onHoverInside;
  final Function? onHoverOutside;

  const TransparentImageButton.assets(this.imagePath,
      {Key? key,
        this.frameBuilder,
        this.semanticLabel,
        this.excludeFromSemantics = false,
        this.scale,
        this.width,
        this.height,
        this.color,
        this.colorBlendMode,
        this.fit,
        this.alignment = Alignment.center,
        this.repeat = ImageRepeat.noRepeat,
        this.centerSlice,
        this.matchTextDirection = false,
        this.gaplessPlayback = false,
        this.package,
        this.filterQuality = FilterQuality.low,
        this.onTapInside,
        this.onTapOutside,
        this.onHoverInside,
        this.onHoverOutside,
        this.updateCursor = true,
        this.offCursor = SystemMouseCursors.basic,
        this.onCursor = SystemMouseCursors.click,
        this.opacityThreshold = 0.0,
        this.checkTap = false})
      : super(key: key);

  // TODO: TransparentImageButton.network

  @override
  State<StatefulWidget> createState() => _TransparentImageButton();

  /// Run against a real-world image, the previous example renders the following
  /// image.
  ///
  /// {@animation 400 400 https://flutter.github.io/assets-for-api-docs/assets/widgets/frame_builder_image.mp4}
  final ImageFrameBuilder? frameBuilder;

  /// If non-null, require the image to have this width.
  ///
  /// If null, the image will pick a size that best preserves its intrinsic
  /// aspect ratio.
  ///
  /// It is strongly recommended that either both the [width] and the [height]
  /// be specified, or that the widget be placed in a context that sets tight
  /// layout constraints, so that the image does not change size as it loads.
  /// Consider using [fit] to adapt the image's rendering to fit the given width
  /// and height if the exact image dimensions are not known in advance.
  final double? width;

  /// If non-null, require the image to have this height.
  ///
  /// If null, the image will pick a size that best preserves its intrinsic
  /// aspect ratio.
  ///
  /// It is strongly recommended that either both the [width] and the [height]
  /// be specified, or that the widget be placed in a context that sets tight
  /// layout constraints, so that the image does not change size as it loads.
  /// Consider using [fit] to adapt the image's rendering to fit the given width
  /// and height if the exact image dimensions are not known in advance.
  final double? height;

  /// If non-null, this color is blended with each image pixel using [colorBlendMode].
  final Color? color;

  /// Used to set the [FilterQuality] of the image.
  ///
  /// Use the [FilterQuality.low] quality setting to scale the image with
  /// bilinear interpolation, or the [FilterQuality.none] which corresponds
  /// to nearest-neighbor.
  final FilterQuality filterQuality;

  /// Used to combine [color] with this image.
  ///
  /// The default is [BlendMode.srcIn]. In terms of the blend mode, [color] is
  /// the source and this image is the destination.
  ///
  /// See also:
  ///
  ///  * [BlendMode], which includes an illustration of the effect of each blend mode.
  final BlendMode? colorBlendMode;

  /// How to inscribe the image into the space allocated during layout.
  ///
  /// The default varies based on the other fields. See the discussion at
  /// [paintImage].
  final BoxFit? fit;

  /// How to align the image within its bounds.
  ///
  /// The alignment aligns the given position in the image to the given position
  /// in the layout bounds. For example, an [Alignment] alignment of (-1.0,
  /// -1.0) aligns the image to the top-left corner of its layout bounds, while an
  /// [Alignment] alignment of (1.0, 1.0) aligns the bottom right of the
  /// image with the bottom right corner of its layout bounds. Similarly, an
  /// alignment of (0.0, 1.0) aligns the bottom middle of the image with the
  /// middle of the bottom edge of its layout bounds.
  ///
  /// To display a subpart of an image, consider using a [CustomPainter] and
  /// [Canvas.drawImageRect].
  ///
  /// If the [alignment] is [TextDirection]-dependent (i.e. if it is a
  /// [AlignmentDirectional]), then an ambient [Directionality] widget
  /// must be in scope.
  ///
  /// Defaults to [Alignment.center].
  ///
  /// See also:
  ///
  ///  * [Alignment], a class with convenient constants typically used to
  ///    specify an [AlignmentGeometry].
  ///  * [AlignmentDirectional], like [Alignment] for specifying alignments
  ///    relative to text direction.
  final AlignmentGeometry alignment;

  /// How to paint any portions of the layout bounds not covered by the image.
  final ImageRepeat repeat;

  /// The center slice for a nine-patch image.
  ///
  /// The region of the image inside the center slice will be stretched both
  /// horizontally and vertically to fit the image into its destination. The
  /// region of the image above and below the center slice will be stretched
  /// only horizontally and the region of the image to the left and right of
  /// the center slice will be stretched only vertically.
  final Rect? centerSlice;

  /// Whether to paint the image in the direction of the [TextDirection].
  ///
  /// If this is true, then in [TextDirection.ltr] contexts, the image will be
  /// drawn with its origin in the top left (the "normal" painting direction for
  /// images); and in [TextDirection.rtl] contexts, the image will be drawn with
  /// a scaling factor of -1 in the horizontal direction so that the origin is
  /// in the top right.
  ///
  /// This is occasionally used with images in right-to-left environments, for
  /// images that were designed for left-to-right locales. Be careful, when
  /// using this, to not flip images with integral shadows, text, or other
  /// effects that will look incorrect when flipped.
  ///
  /// If this is true, there must be an ambient [Directionality] widget in
  /// scope.
  final bool matchTextDirection;

  /// Whether to continue showing the old image (true), or briefly show nothing
  /// (false), when the image provider changes.
  final bool gaplessPlayback;

  /// A Semantic description of the image.
  ///
  /// Used to provide a description of the image to TalkBack on Android, and
  /// VoiceOver on iOS.
  final String? semanticLabel;

  /// Whether to exclude this image from semantics.
  ///
  /// Useful for images which do not contribute meaningful information to an
  /// application.
  final bool excludeFromSemantics;

  /// Whether to show a different mouse cursor while hovering over the image (true),
  /// or prevent the cursor from changing at all while hovering over the image (false).
  final bool updateCursor;

  /// Used to set the cursor type while hovering over an opaque portion of the image.
  ///
  /// Defaults to [SystemMouseCursors.click].
  ///
  /// If [updateCursor] is set to false, this will have no effect.
  final SystemMouseCursor onCursor;

  /// Used to set the cursor type while hovering over a transparent portion of the image.
  ///
  /// Defaults to [SystemMouseCursors.basic].
  ///
  /// If [updateCursor] is set to false, this will have no effect.
  final SystemMouseCursor offCursor;

  /// How opaque a pixel must to be before being recognized as part of the image. This
  /// is useful for images that have a partially-transparent border that shouldn't be clickable.
  final double opacityThreshold;

  /// Whether to check touches on a full tap (true), or onPanDown (false).
  ///
  /// For applications where you might be panning or scrolling, it's recommended to
  /// set this to true in order to avoid firing when the user is tring to scroll.
  ///
  /// Defaults to false.
  final bool checkTap;

  // Package
  final String? package;

  // Scale
  final double? scale;
}

class _TransparentImageButton extends State<TransparentImageButton> {
  GlobalKey imageKey = GlobalKey();

  img.Image? photo;

  SystemMouseCursor cursor = SystemMouseCursors.basic;

  @override
  Widget build(BuildContext context) {
    return MouseRegion(
      cursor: cursor,
      onHover: (details) async {
        // Check if the current pixel is transparent
        var search = await searchPixel(details.position, true);

        if (widget.updateCursor == true) {
          if (search == true) {
            cursor = widget.onCursor;
            setState(() {});
          } else {
            cursor = widget.offCursor;
            setState(() {});
          }
        }
      },
      onExit: (event) {
        // Make sure the cursor is properly set when exiting
        if (widget.updateCursor == true) {
          cursor = widget.offCursor;
          setState(() {});
        }
      },
      child: GestureDetector(
        onPanDown: (details) => widget.checkTap ? null : searchPixel(details.globalPosition, false),
        onTapDown: (details) => widget.checkTap ? searchPixel(details.globalPosition, false) : null,
        // onPanUpdate: (details) => searchPixel(details.globalPosition),
        child: Image.asset(
          widget.imagePath,
          key: imageKey,
          frameBuilder: widget.frameBuilder,
          semanticLabel: widget.semanticLabel,
          excludeFromSemantics: widget.excludeFromSemantics,
          scale: widget.scale,
          width: widget.width,
          height: widget.height,
          color: widget.color,
          colorBlendMode: widget.colorBlendMode,
          fit: widget.fit,
          alignment: widget.alignment,
          repeat: widget.repeat,
          centerSlice: widget.centerSlice,
          matchTextDirection: widget.matchTextDirection,
          gaplessPlayback: widget.gaplessPlayback,
          package: widget.package,
          filterQuality: widget.filterQuality,
        ),
      ),
    );
  }

  dynamic searchPixel(Offset globalPosition, bool hover) async {
    if (photo == null) {
      await loadImageBundleBytes();
    }
    return _calculatePixel(globalPosition, hover);
  }

  bool _calculatePixel(Offset globalPosition, bool hover) {
    RenderBox box = imageKey.currentContext!.findRenderObject() as RenderBox;
    Offset localPosition = box.globalToLocal(globalPosition);

    double px = localPosition.dx;
    double py = localPosition.dy;

    double widgetScale = box.size.width / photo!.width;
    px = (px / widgetScale);
    py = (py / widgetScale);

    int pixel32 = photo!.getPixelSafe(px.toInt(), py.toInt());
    int hex = abgrToArgb(pixel32);

    if (Color(hex).opacity <=
        widget.opacityThreshold) { // Pixel meets the opacity threshold
      if (hover && widget.onHoverOutside != null) {
        widget.onHoverOutside!();
      } else if (!hover && widget.onTapOutside != null) {
        widget.onTapOutside!();
      }
    } else { // Pixel is opaque
      if (hover) { // This is a hover event, not a tap
        if (widget.onHoverInside != null) {
          widget.onHoverInside!();
        }

        return true;
      } else if (widget.onTapInside != null) {
        widget.onTapInside!();
      }
    }

    return false;
  }

  Future<void> loadImageBundleBytes() async {
    ByteData imageBytes = await rootBundle.load(widget.imagePath);
    setImageBytes(imageBytes);
  }

  void setImageBytes(ByteData imageBytes) {
    List<int> values = imageBytes.buffer.asUint8List();
    photo = null;
    photo = img.decodeImage(values)!;
  }
}

// image lib uses uses KML color format, convert #AABBGGRR to regular #AARRGGBB
int abgrToArgb(int argbColor) {
  int r = (argbColor >> 16) & 0xFF;
  int b = argbColor & 0xFF;
  return (argbColor & 0xFF00FF00) | (b << 16) | r;
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:transparent_image_button/transparent_image_button.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Transparent Image Button Ex',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Transparent Image Button Ex'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            TransparentImageButton.assets(
              "assets/images/egg.png",
              width: 200,
              opacityThreshold: 0.8,
              onTapInside: () => print("You tapped the image."),
              onTapOutside: () => print("You tapped outside the image."),
              onHoverInside: () => print("You're hovering inside the image."),
              onHoverOutside: () => print("You're hovering outside the image."),
            ),
          ],
        ),
      ),
    );
  }
}
